packages/e2e/src/tools/multi-delegation-data-gen/utils/utils.ts

```diff
@@ -0,0 +1,438 @@
+/* eslint-disable no-console, max-statements, max-params, @typescript-eslint/no-floating-promises */
+import { ValueTransferConfig, configLoader } from './config';
+
+import { AddressType, AsyncKeyAgent, GroupedAddress, util } from '@cardano-sdk/key-management';
+import { Cardano } from '@cardano-sdk/core';
+import { Files, Paths } from './files';
+import {
+  KeyAgentFactoryProps,
+  MINUTE,
+  firstValueFromTimed,
+  getEnv,
+  getWallet,
+  submitAndConfirm,
+  walletReady,
+  walletVariables
+} from '../../../';
+import { Observable, filter, firstValueFrom, map } from 'rxjs';
+import { PersonalWallet } from '@cardano-sdk/wallet';
+import { TaskResult, TerminalProgressMonitor } from './terminal-progress-monitor';
+import { logger } from '@cardano-sdk/util-dev';
+import chalk from 'chalk';
+
+/**
+ * Gets grouped addresses with increasing stake key index.
+ *
+ * @param keyAgent The key agent to be used to generate the addresses.
+ * @param addressCount Number of addresses to be generated.
+ */
+export const getStakeAddresses = async (
+  keyAgent: AsyncKeyAgent,
+  addressCount: number
+): Promise<Array<GroupedAddress>> => {
+  const addresses = new Array<GroupedAddress>();
+
+  for (let i = 0; i < addressCount; ++i) {
+    const address = await keyAgent.deriveAddress(
+      {
+        index: 0,
+        type: AddressType.External
+      },
+      i
+    );
+
+    addresses.push(address);
+  }
+
+  return addresses;
+};
+
+/**
+ * Create stake key registration certificate.
+ *
+ * @param rewardAccount The reward account to register to.
+ */
+const createStakeKeyRegistrationCert = (rewardAccount: Cardano.RewardAccount): Cardano.Certificate => ({
+  __typename: Cardano.CertificateType.StakeKeyRegistration,
+  stakeKeyHash: Cardano.RewardAccount.toHash(rewardAccount)
+});
+
+/**
+ * Creates a delegation certificate for the given reward account and pool id.
+ *
+ * @param rewardAccount The reward account that is delegating.
+ * @param poolId The pool id of the pool we are delegating to.
+ */
+const createDelegationCert = (rewardAccount: Cardano.RewardAccount, poolId: Cardano.PoolId): Cardano.Certificate => ({
```

Comment:
This and `createStakeKeyRegistrationCert` seem like they should be part of a more generic utility, maybe within `tx-construction` :thinking:

Reply:
Delegate certificate is [already there](https://github.com/input-output-hk/cardano-js-sdk/blob/master/packages/tx-construction/src/tx-builder/TxBuilder.ts#LL290C32-L290C32), but it's internal private

Reply:
Yeah, even better